DEV: Apply Ember Octane Upgrades

keegangeorge opened the pull request:
This PR applies some ember upgrades to the plugin including things such as convert to native JS classes, apply angle bracket syntax, colocate components, use tracked getters/setters rather than computed properties, etc.

_Note this is the first set of changes (revolving around `/activity/reactions` `/notifications/reactions-received` routes), next set of changes (separate PR) will focus on updates to the widgets._

keegangeorge committed:
DEV: Apply Ember Octane Upgrades

Convert to native JS classes, apply angle bracket syntax, etc.

keegangeorge commented on assets/javascripts/discourse/components/discourse-reactions-reaction-post.hbs:
```diff
@@ -0,0 +1,59 @@
+<div
+  class="user-stream-item item {{if this.moderatorAction 'moderatorAction'}}"
```
Unsure if converting the `classNameBindings` to this is the correct usage here.

davidtaylorhq replied:
Yeah since there's only one conditional class, I think this is sensible. If the logic was mega-complex then it could've been refactored into getter, but there's no need for that here

jjaffeux replied:
you can use `{{concat-class "user-stream-item item" (if this.moderatorAction "moderatorAction")}}`

It avoids all the handling of white space

davidtaylorhq commented on assets/javascripts/discourse/routes/user-activity-reactions.js:
```diff
@@ -1,13 +1,13 @@
+import Route from "@ember/routing/route";
 import CustomReaction from "../models/discourse-reactions-custom-reaction";
 
-export default DiscourseRoute.extend({
```
I think it's probably safer to stick to `DiscourseRoute` - it has a few helpers that we use. You can still use the class-based syntax though 👌

davidtaylorhq commented on assets/javascripts/discourse/controllers/user-activity-reactions.js:
```diff
@@ -1,20 +1,22 @@
+  canLoadMore = true;
+  loading = false;
+  beforeLikeId = null;
+  beforeReactionUserId = null;
```
I suspect these will need to be `@tracked` so that setting them via `this.blah = ` actually triggers a rerender.

To make sure things are working, I'd recommend setting up your local environment with more than one page (25?) reactions, and check that infinite scrolling and the loading-spinner are working.

keegangeorge replied:
Interesting, yes it wasn't working correctly. I was getting new results loaded but the loading indicator wasn't appearing till I made the properties `@tracked`

davidtaylorhq commented on assets/javascripts/discourse/controllers/user-activity-reactions.js:
```diff
@@ -61,18 +63,18 @@ export default Controller.extend({
-  },
+  }
 
   @observes("canLoadMore")
```
Can you confirm that this is working correctly? I'm not sure whether `@observes` works under octane `@tracked`/autotracking

keegangeorge replied:
Yup it wasn't working correctly, I've removed it in the latest commit and am updating `showFooter` differently, thanks!

davidtaylorhq commented on assets/javascripts/discourse/controllers/user-activity-reactions.js:
```diff
@@ -61,18 +63,18 @@ export default Controller.extend({
 
   @observes("canLoadMore")
   _showFooter() {
-    this.set("application.showFooter", !this.canLoadMore);
```
`ApplicationController` hasn't been converted to use `@tracked` yet, so I think we've gotta stick with `.set()` here for now

davidtaylorhq reviewed (commented):
Looks great - nice work @keegangeorge!

We'll need to add a `.discourse-compatibility` entry to make sure that these new changes are not installed on `stable` versions of Discourse. (see [user-notes](https://github.com/discourse/discourse-user-notes/blob/main/.discourse-compatibility) for an example)

keegangeorge committed:
DEV: Improve class whitespace handling

keegangeorge committed:
DEV: Use `DiscourseRoute` instead of `Route`

keegangeorge committed:
DEV: Make controller properties `@tracked`

keegangeorge committed:
FIX: Remove not functioning observer

Removes the observer and sets the footer property by default in the constructor and updates the property through the `loadMore()` action

keegangeorge commented:
Pushed some updates based on the review and feedback. May I have another look over please :)

keegangeorge committed:
DEV: Update `.discourse-compatibility`

Glimmer updates will only function correctly on a more recent version of Discourse.

davidtaylorhq commented on assets/javascripts/discourse/controllers/user-activity-reactions.js:
```diff
@@ -31,24 +38,25 @@ export default Controller.extend({
   @action
   loadMore() {
     if (!this.canLoadMore || this.loading || !this.reactionsUrl) {
+      this.set("application.showFooter", !this.canLoadMore);
```
Much cleaner than the `@observes` approach, nice! 🙌

davidtaylorhq reviewed (approved):
Looks great, thanks @keegangeorge!
